cpu/stm32/periph/sdmmc.c

```diff
@@ -0,0 +1,849 @@
+    sdmmc->DTIMER = xfer->write ? SDMMC_DATA_W_TIMEOUT : SDMMC_DATA_R_TIMEOUT;
+    sdmmc->DLEN = xfer->block_size * xfer->block_num;
+
+#if IS_USED(MODULE_PERIPH_DMA)
+    const sdmmc_conf_t *conf = dev->config;
+
+    if (_use_dma(conf)) {
+        dma_acquire(conf->dma);
+        dma_setup(conf->dma, conf->dma_chan, (uint32_t*)&(conf->dev->FIFO),
+                  xfer->write ? DMA_MEM_TO_PERIPH : DMA_PERIPH_TO_MEM,
+                  DMA_DATA_WIDTH_WORD, false);
+        /* additional configuration needed for F2/F4/F7 */
+        dma_setup_ext(conf->dma, DMA_BURST_INCR4, DMA_BURST_INCR4,
+                      true, DMA_FIFO_FULL, true);
+    }
+#endif
+
+    return 0;
+}
+
+static int _xfer_execute(sdmmc_dev_t *sdmmc_dev, sdmmc_xfer_desc_t *xfer,
+                         const void *data_wr, void *data_rd,
+                         uint16_t *done)
+{
+    assert(xfer);
+    assert((xfer->write && data_wr) || (!xfer->write && data_rd));
+
+    /* check the alignment required for the buffers */
+    assert(HAS_ALIGNMENT_OF(data_wr, SDMMC_CPU_DMA_ALIGNMENT));
+    assert(HAS_ALIGNMENT_OF(data_rd, SDMMC_CPU_DMA_ALIGNMENT));
+
+    stm32_sdmmc_dev_t *dev = container_of(sdmmc_dev, stm32_sdmmc_dev_t, sdmmc_dev);
+    assert(dev);
+
+    const sdmmc_conf_t *conf = dev->config;
+    assert(conf);
+
+    SDMMC_TypeDef *sdmmc = conf->dev;
+
+    uint32_t *data_to_read = data_rd;
+    const uint32_t *data_to_write = data_wr;
+
+    /* clear data transfer related static interrupt flags */
+    sdmmc->ICR = SDMMC_INT_STATIC_DATA;
+
+    /* enable used data interrupts */
+    uint32_t status_mask = SDMMC_MASK_DCRCFAILIE | SDMMC_MASK_DTIMEOUTIE;
+    status_mask |= xfer->write ? SDMMC_MASK_TXUNDERRIE : SDMMC_MASK_RXOVERRIE;
+    status_mask |= (xfer->block_num == 1) ? SDMMC_MASK_DBCKENDIE
+                                          : SDMMC_MASK_DATAENDIE;
+#if defined(SDMMC_MASK_STBITERRIE)
+    /* Some CMSIS headers define the STBITERR enable and clear flag but not
+     * the status flag. For such MCU models, the STBITERR flag is not documented
+     * in the RM but marked as reserved. */
+    status_mask |= SDMMC_MASK_STBITERRIE;
+#endif
+
+    sdmmc->MASK |= status_mask;
+
+    /* configure data transfer block size, read direction and enable the transfer */
+    uint32_t dctrl = 0;
+    dctrl |= bitarithm_msb(xfer->block_size) << SDMMC_DCTRL_DBLOCKSIZE_Pos;
+    dctrl |= !xfer->write ? SDMMC_DCTRL_DTDIR : 0;
+
+    /* TODO: at the moment only 32-bit word transfers supported */
+    assert((xfer->block_size % sizeof(uint32_t)) == 0);
+    uint32_t fifo_words = sdmmc->DLEN >> 2; /* 32-bit words to read or write */
+
+    /* read data blocks from or write data blocks to the card */
+    if (_use_dma(conf)) {
+#if IS_USED(MODULE_PERIPH_DMA)
+        dma_prepare(conf->dma,
+                    (xfer->write) ? (void *)data_to_write : data_to_read,
+                    fifo_words, 1);
+        dma_start(conf->dma);
+
+        /* TODO SDMMC_IDMA_IDMAEN */
```

Comment:
?

Reply:
The SDMMC of L4+ and some L5 MCUs have an internal DMA conntroller (IDMA) instead of using the DMA peripheral. We need a separate handling for it. `_use_dma` just returns false if `SDMMC_IDMA_IDMAEN` is defined so that this part should be optimized out by the compiler.

Reply:
Ah ok - then please add that information to the comment.